File: backend/app/models/item.rb
# frozen_string_literal: true

class Item < ApplicationRecord
  belongs_to :user
  acts_as_taggable_on :tags
  acts_as_taggable_on :categories

  validates :title, presence: true
  validates :url,
            presence: true,
            uniqueness: {
              scope: %i[user url],
              message: 'already exists in collection'
            },
            format: {
              with: URI::DEFAULT_PARSER.make_regexp,
              message: 'invalid URL'
            }

  validate do |item|
    collector = WebCollector.new(item.url)

    if collector.valid?
      collected = collector.collect
      item.mediatype = collected.fetch(:mediatype, 'invalid')
      item.size = collected[:size]
      item.category_list = collected[:categories]
    else
      item.errors[:url] << 'Invalid url'
    end
  end
end

File: frontend/models/Item.ts
export interface Item {
  id: number
  title: string
  url: string
  uid: number
  tags: string[]
  categories: string[]
  mediatype: string
  size: number
}

export interface IItemJson {
  id: number
  title: string
  url: string
  categories: { name: string; taggings_count: number }[]
  created_at: string
  mediatype: string
  size: number
  tags: { name: string; taggings_count: number }[]
  updated_at: string
}
